pl-faded-parsons.py

```diff
@@ -1,314 +1,538 @@
+from typing import Union, List, Dict, Literal, Any, get_args, get_origin, cast
+from datetime import datetime
+from enum import Enum
+
+UnionType = Union # replace with an import when python>=3.10
+NoneType = type(None) # replace with an import when python>=3.10
 
 #
-# Helper functions
+# Common Interfaces for Parsing/Generating Data
 #
-def read_file_lines(data, filename, error_if_not_found=True):
-    """Return a string of newline-separated lines of code from some file in serverFilesQuestion."""
-    path = os.path.join(data["options"]["question_path"], 'serverFilesQuestion', filename)
-    try:
-        f = open(path, 'r')
-        return f.read()
-    except FileNotFoundError as e:
-        if error_if_not_found:
-            raise e
-        else:
-            return False
-
-
-def get_answers_name(element_html):
-    # use answers-name to namespace multiple pl-faded-parsons elements on a page
-    element = xml.fragment_fromstring(element_html)
-    return pl.get_string_attrib(element, 'answers-name', '')
-
-
-def get_student_code(element_html, data):
-    answers_name = get_answers_name(element_html)
-    student_code = data['submitted_answers'].get(answers_name + 'student-parsons-solution', None)
-    return student_code
-
-
-def parse_lines(language, lines):
-    'Reads lines structured by pl-faded-parsons.js codelineSummary'
-    for line in lines:
-        old_segments = line['segments']['givenSegments']
-        segments = [{ "code" : { "content" : old_segments[0] }}]
-
-        for segment, fill in zip(old_segments[1:], line['segments']['blankValues']):
-            segments.append({ "blank" : { "default" : fill    }})
-            segments.append({ "code"  : { "content" : segment }})
-
-        yield {
-            "language" : language,
-            "segments" : segments,
-            "indent": line.get('indent', 0)
-        }
-
-
-def load_previous_state(language, old_starter, old_submission, indent_size=4):
-    scrambled = list(parse_lines(language, old_starter))
-    given = []
-
-    for data in parse_lines(language, old_submission):
-        data['indent'] *= indent_size
-        given.append(data)
-
-    return scrambled, given
-
-
-def load_starter_and_given(raw_lines, language, indent_size=4, max_distractors=10):
-    line_segments = [ line.strip().split('!BLANK') for line in raw_lines.strip().split('\n') ]
-
-    scrambled = []
-    given = []
-    distractors = []
-
-    for segments in line_segments:
-        new_line = { "language": language }
 
-        matches = re.findall(r'#blank [^#]*', segments[-1])
-        tail = re.sub(r'#blank [^#]*', '', segments[-1])
-        blank_count = len(segments) - 1
-        fills = list(map(lambda e: e.replace('#blank ', ""), matches)) + [""] * (blank_count-len(matches))
-        segments[-1] = tail
+"""
+We use dataclasses instead of `TypeDict`s to allow for type checking in constructors
+"""
+
+@dataclass(frozen=True)
```

Comment:
Every call of `dataclass` should use `slots=True`, particularly when `frozen=True` -- massive perf gain during construction, slight hit during mutation (not an issue if frozen).